// vidstack/dist/prod/providers/video/presentation/video-presentation.js
import { listenEvent, DOMEvent } from 'maverick.js/std';
import { $ as canUseVideoPresentation } from '../../../media-core.js';

class VideoPresentation {
  constructor(_video, _media) {
    this.ta = _video;
    this.n = _media;
    listenEvent(this.ta, "webkitpresentationmodechanged", this.Kh.bind(this));
  }
  Oa = "inline";
  get Jh() {
    return canUseVideoPresentation(this.ta);
  }
  async Ih(mode) {
    if (this.Oa === mode)
      return;
    this.ta.webkitSetPresentationMode(mode);
  }
  Kh() {
    const prevMode = this.Oa;
    this.Oa = this.ta.webkitPresentationMode;
    this.n.player?.dispatchEvent(
      new DOMEvent("video-presentation-change", {
        detail: this.Oa,
        trigger: event
      })
    );
    ["fullscreen", "picture-in-picture"].forEach((type) => {
      if (this.Oa === type || prevMode === type) {
        this.n.delegate.U(`${type}-change`, {
          detail: this.Oa === type,
          trigger: event
        });
      }
    });
  }
}
class FullscreenPresentationAdapter {
  constructor(_presentation) {
    this.Hh = _presentation;
  }
  get active() {
    return this.Hh.Oa === "fullscreen";
  }
  get supported() {
    return this.Hh.Jh;
  }
  async enter() {
    this.Hh.Ih("fullscreen");
  }
  async exit() {
    this.Hh.Ih("inline");
  }
}
class PIPPresentationAdapter {
  constructor(_presentation) {
    this.Hh = _presentation;
  }
  get active() {
    return this.Hh.Oa === "picture-in-picture";
  }
  get supported() {
    return this.Hh.Jh;
  }
  async enter() {
    this.Hh.Ih("picture-in-picture");
  }
  async exit() {
    this.Hh.Ih("inline");
  }
}

export { FullscreenPresentationAdapter as F, PIPPresentationAdapter as P, VideoPresentation as V };
