// vuepress-theme-hope/lib/client/modules/info/components/WordInfo.js
import { defineComponent, h } from "vue";
import { WordIcon } from "@theme-hope/modules/info/components/icons";
import { useMetaLocale } from "@theme-hope/modules/info/composables/index";
export default defineComponent({
    name: "ReadTimeInfo",
    inheritAttrs: false,
    props: {
        /**
         * Reading time information
         *
         * é˜…è¯»æ—¶é—´ä¿¡æ¯
         */
        readingTime: {
            type: Object,
            default: () => null,
        },
        /**
         * Reading time locale
         *
         * é˜…è¯»æ—¶é—´è¯­è¨€çŽ¯å¢ƒ
         */
        readingTimeLocale: {
            type: Object,
            default: () => null,
        },
        /**
         * Whether in pure mode
         *
         * æ˜¯å¦å¤„äºŽçº¯å‡€æ¨¡å¼
         */
        pure: Boolean,
    },
    setup(props) {
        const metaLocale = useMetaLocale();
        return () => props.readingTimeLocale?.words
            ? h("span", {
                class: "page-word-info",
                "aria-label": `${metaLocale.value.words}${props.pure ? "" : "ðŸ” "}`,
                ...(props.pure ? {} : { "data-balloon-pos": "down" }),
            }, [
                h(WordIcon),
                h("span", props.readingTimeLocale?.words),
                h("meta", {
                    property: "wordCount",
                    content: props.readingTime?.words,
                }),
            ])
            : null;
    },
});
