// vidstack/dist/prod/providers/html/native-audio-tracks.js
import { listenEvent } from 'maverick.js/std';
import { N as LIST_ADD, W as LIST_REMOVE, L as LIST_SELECT } from '../../media-core.js';

class NativeAudioTracks {
  constructor(_provider, _context) {
    this.m = _provider;
    this.ph = _context;
    this.Nh.onaddtrack = this.Ph.bind(this);
    this.Nh.onremovetrack = this.Qh.bind(this);
    this.Nh.onchange = this.Rh.bind(this);
    listenEvent(this.ph.audioTracks, "change", this.Sh.bind(this));
  }
  get Nh() {
    return this.m.media.audioTracks;
  }
  Ph(event) {
    const _track = event.track;
    if (_track.label === "")
      return;
    const audioTrack = {
      id: _track.id + "",
      label: _track.label,
      language: _track.language,
      kind: _track.kind,
      selected: false
    };
    this.ph.audioTracks[LIST_ADD](audioTrack, event);
    if (_track.enabled)
      audioTrack.selected = true;
  }
  Qh(event) {
    const track = this.ph.audioTracks.getById(event.track.id);
    if (track)
      this.ph.audioTracks[LIST_REMOVE](track, event);
  }
  Rh(event) {
    let enabledTrack = this.Oh();
    if (!enabledTrack)
      return;
    const track = this.ph.audioTracks.getById(enabledTrack.id);
    if (track)
      this.ph.audioTracks[LIST_SELECT](track, true, event);
  }
  Oh() {
    return Array.from(this.Nh).find((track) => track.enabled);
  }
  Sh(event) {
    const { current } = event.detail;
    if (!current)
      return;
    const track = this.Nh.getTrackById(current.id);
    if (track) {
      const prev = this.Oh();
      if (prev)
        prev.enabled = false;
      track.enabled = true;
    }
  }
}

export { NativeAudioTracks as N };
